qvt102: move to a device and add to the rs232 bus

68bit opened the pull request:
This is a proposal to move the QVT102 terminal emulator to the RS232 bus interface, so that it can be used as an optional terminal for a range vintage machine emulators. This proposal only adds it as an option for the SWTPC but I have other machines in mind too.

The QVT102 terminal is well emulated, it uses components appropriate for the late 70's and early 80's, and it uses them in a relatively stock manner, it has an 8k ROM, it's a light weight terminal, and it has 'personalities' or compatibility for some common terminals for this period, such as the ADM3A and TVI910, plus a convenient onscreen setup menu with nvram storage.

It's a 'magic cookie' vintage terminal that needs spaces to change character attributes, and these were so frustrating to use that they are probably unusable - a RAM limitation. However it can do highlighting using the high bit without these.

There are obviously far more capable terminals but some seem disproportionate to the host systems!

cuavas commented on src/devices/bus/rs232/qvt102.cpp:
```diff
@@ -0,0 +1,76 @@
+	PORT_START("FLOW_CONTROL")
+	PORT_CONFNAME(0xf, 1, "Flow control")
+	PORT_CONFSETTING(0x00, "None")
+	PORT_CONFSETTING(0x01, "Terminal DTR to remote CTS")
```
You need to attach a `PORT_CHANGED_MEMBER` to this so you can switch the output state if someone changes this while the emulation is running.

cuavas commented on src/devices/bus/ss50/mps.cpp:
```diff
@@ -48,26 +50,34 @@ class ss50_mps_device : public device_t, public ss50_card_interface
+	PORT_START("CTS_ROUTE")
+	PORT_CONFNAME(1, 0, "CTS route")
+	PORT_CONFSETTING(0, "Wired low (standard)")
+	PORT_CONFSETTING(1, "Connected through (modified)")
```
This should have a `PORT_CHANGED_MEMBER` as well for doing live changes.

68bit replied:
These don't a `PORT_CHANGED_MEMBER` as the port is re-read each time a decision is made. I had been pondering if that was too heavy, and if a separate variable is needed, and then the `PORT_CHANGED_MEMBER` would be needed to update that state. For the control lines, their state is not expected to be changing at a high frequency, so this seems ok. What do you think?

cuavas replied:
You need a port changed member because you're not going to immediately get an update from the MCU when the switch is changed.

When the switch is switched to route, you need to update the output from the fixed value to whatever the terminal last output.  Otherwise it won't update until the next time the terminal changes its output.

When the switch is changed to don't route, you need to update the output from whatever the terminal last output to the fixed value.  Otherwise you'll never update it because you ignore the terminal's output when the switch is set not to route.

68bit replied:
Great point, thanks. Let me rework that.

cuavas commented on src/devices/machine/qvt102.cpp:
```diff
@@ -0,0 +1,692 @@
+
+	if (BIT(m_kbd_bus, 4) == 0) state = BIT(m_keys_special->read(), 0);
+	if (BIT(m_kbd_bus, 5) == 0) state = BIT(m_keys_special->read(), 1);
+	if (BIT(m_kbd_bus, 6) == 0) state = BIT(m_keys_special->read(), 2);
```
Does this really have a priority mechanism?  If it doesn't, it should be `state &=` on each line; if it really does have a priority mechanism it should be `else if` to make it clearer.

68bit replied:
This and other some other queries are for the old code, but let me take a look.

cuavas commented on src/devices/machine/qvt102.cpp:
```diff
@@ -0,0 +1,692 @@
+			state = BIT(m_keys_p1[(m_kbd_bus >> 4) & 7]->read(), i);
+		if (BIT(m_kbd_p2, i) == 0)
+			state = BIT(m_keys_p2[(m_kbd_bus >> 4) & 7]->read(), i);
+	}
```
This should use `state &=` as well.

cuavas commented on src/devices/machine/qvt102.cpp:
```diff
@@ -0,0 +1,692 @@
+	{
+		if (BIT(m_jumper->read(), 2))
+		{
+			fprintf(stderr, "######## rs232_write_dtr %02x\n", BIT(data, 4));
```
Please use `logerror` for general diagnostic messages, and don't put the hashes.  The device tag will autmatically be logged that way, it will be visible to the debugger, and you can send it to the console with `-oslog`.

Why are you logging a single bit as a two-digit hex number?

Also, this should probably use `LOG` from `logmacro.h` to put it behind the `VERBOSE` macro as it isn't really something unusual.

cuavas commented on src/devices/machine/qvt102.cpp:
```diff
@@ -0,0 +1,692 @@
+		if (BIT(m_latch, 0) == 0)
+			m_aux->write_txd(state);
+
+		fprintf(stderr, "rs232_write_txd %02x\n", state);
```
Once again, use MAME's diagnostic outputs appropriately, and put stuff that shouldn't be surprising behind `VERBOSE` using `logmacro.h` stuff.

cuavas commented on src/devices/machine/qvt102.cpp:
```diff
@@ -0,0 +1,692 @@
+	{
+		if (BIT(m_jumper->read(), 3))
+		{
+			fprintf(stderr, "######## rs232_write_dtr %02x\n", state);
```
Same comments as other diagnostics.

cuavas commented on src/devices/machine/qvt102.cpp:
```diff
@@ -0,0 +1,692 @@
+
+WRITE_LINE_MEMBER(qvt102_device::rs232_conn_rxd_w)
+{
+	fprintf(stderr, "qvt102_device::rs232_rxd_w %02x\n", state);
```
Same comment as before.

cuavas commented on src/devices/machine/qvt102.h:
```diff
@@ -0,0 +1,135 @@
+	DECLARE_WRITE_LINE_MEMBER(rs232_conn_dsr_w);
+	DECLARE_WRITE_LINE_MEMBER(rs232_conn_ri_w);
+	DECLARE_WRITE_LINE_MEMBER(rs232_conn_cts_w);
+	DECLARE_WRITE_LINE_MEMBER(rs232_conn_rxd_w);
```
Please clearly separate configuration member functions and live inputs.  Often configuration members are declared above the constructor to make things clearer.

68bit replied:
Could I leave these grouped together here just for a bit. I have in mind adding a new class defining these, so that the terminals have a common interface to the rs232 bus, and it will be quicker to clean up with them all together. This class might implement some common null modem options too, too refactor this code. Looking forward some terminals have more than one host interface, such as the tv900 and it would be nice to be able to route both to host rs232 ports - this area needs some 'architecture' design that might be better driven by other contributors.

cuavas replied:
You should at least separate them and mark which ones are for configuration and which ones are inputs.  Also, the QVT-102 doesn't have an RI input as far as I can tell, and it has a clock output.

cuavas commented on src/devices/machine/qvt102.cpp:
```diff
@@ -0,0 +1,692 @@
+{
+	m_latch = 0;
+	m_kbd_data = 0;
+	m_kbd_bus = m_kbd_p1 = m_kbd_p2 = 0xff;
```
Are these actual latches wired to the machine's system reset?  If not, you should only initialise them in `device_start` and allow the CPUs/MCUs to set them after they're reset.

cuavas commented on src/devices/machine/qvt102.cpp:
```diff
@@ -0,0 +1,692 @@
+	m_maincpu->set_addrmap(AS_PROGRAM, &qvt102_device::mem_map);
+
+	// needs a tight sync with the keyboard cpu
+	config.m_perfect_cpu_quantum = subtag("maincpu");
```
This should only ever be done in top-level machine configurations.  If it's allowed to be done from something that can be instantiated in a slot device, it could cause drivers to break in weird ways.  You'll have to find some other way to deal with synchronisation issues.

68bit commented:
Thanks for the review again. The debug prints were included by mistake sorry, all removed. The bulk of the code is simply moved over so not new code and while I am happy to study any feedback, and will, might it be possible to just review the changes?

cuavas commented:
That would be a lot easier if you'd actually used `git mv` to move the code so the parts that were changed to turn it into a device were more obvious.

At any rate, now that it's a device that can be instantiated in a slot, you need to find a different way to deal with synchronisation than setting perfect sync on a CPU.  That's going to cause problems, and it's part of turning it into a device.

The changes to add CTS routing are also yours.

I think the majority of the comments actually do apply to the changes.

startaq commented:
While I agree that it would be neat to have the qvt102 available as an rs232 bus option, it really needs to keep being usable standalone too. For example, I want to start the driver and attach it's port to a TCP port on the host system and send data to it. This would be impossible if it's only a bus device. Also the machine is usable standalone too: You can write text and send it to the printer port.

So as driver author this gets a NAK from me until a standalone option is added.

68bit commented:
> So as driver author this gets a NAK from me until a standalone option is added.

Then you will be glad to know that this PR does include the option for a stand alone machine, there are two front ends now, and I have tested this with the null modem bit banger running a shell on a host and then running the ncurses tests etc and it looks good, this emulator looks in good shape, thanks.

Btw: I am looking into the timing issue, and see that the main cpu irq uses the bit banding to read from the keyboard. It works for me without the `m_perfect_cpu_quantum` but since the timing is to tight I am looking into the options and perhaps using `boost_interleave()` on edges of the line. Still exploring.

startaq commented:
Thank you for the clarification, I was still on my mobile and didn't have time to take a detailed look.

The timing issue manifests itself as missing keys when typing quickly. It might be possible to solve it another way with explicit calls to synchronize.

68bit commented:
Much of the feedback has been addressed. Added a 'git mv' - squash it if you wish. Had another go at the CTS routing, to update it on a config change. Looked into the timing issue, added a note on the timing, and tried a workaround using `boost_interleave()`, but had not seen and keyboard loss. Looked into the logical-and of some of the bits, and they were being scanned separately, but your suggest seems plausible and is implemented. Have not revised the reset ops, or reordered the interface functions.

68bit commented:
Rebased, and added the CTS route input changed code to mps.cpp thanks.

68bit commented:
Rebased.

Made the qtv102 a default rs232 bus device, like the swtpc 8212 has been.

Tried the new DCE-DCE bridge and that word well for the QVT102 and the SWTPC 8212, thanks. So removed stand alone QVT102 front end from this PR as the bridge can be used.

Presume the IE15 and SWTPC 8212 stand alone front ends can also be removed now too?

68bit commented:
And had to remove `qvt102.cpp` from `mess.flt`, and added `dcebridge.cpp`.

cuavas commented on src/devices/machine/qvt102.cpp:
```diff
@@ -347,53 +292,74 @@ INPUT_PORTS_END
+
+	if (BIT(m_kbd_bus, 4) == 0) state &= BIT(m_keys_special->read(), 0);
+	if (BIT(m_kbd_bus, 5) == 0) state &= BIT(m_keys_special->read(), 1);
+	if (BIT(m_kbd_bus, 6) == 0) state &= BIT(m_keys_special->read(), 2);
```
According to schematics, SHIFT, CTRL and CAPS are the only things connected here.  Perhaps bit 3 is leftover test code, or something that's present on different hardware?  Also, schematics confirm there's no priority logic, and the `&=` is correct here.

cuavas commented on src/devices/machine/qvt102.cpp:
```diff
@@ -347,53 +292,74 @@ INPUT_PORTS_END
+			state &= BIT(m_keys_p1[(m_kbd_bus >> 4) & 7]->read(), i);
 		if (BIT(m_kbd_p2, i) == 0)
-			state = BIT(m_keys_p2[(m_kbd_bus >> 4) & 7]->read(), i);
+			state &= BIT(m_keys_p2[(m_kbd_bus >> 4) & 7]->read(), i);
```
Schematics confirm that `&=` is correct behaviour here as well.

cuavas commented on src/devices/bus/rs232/qvt102.cpp:
```diff
@@ -0,0 +1,87 @@
+	output_dcd(0);
+	output_dsr(0);
+	if (!m_flow_control->read())
+		output_cts(0);
```
This thing needs to be completely dumb - this can get out of sync with device state.  Don't push constant values here, let the devices on the other sides of the proxy push through their correct reset values.

68bit commented:
> I'd prefer if a qvt102 machine was (re-)added. These terminals are DTE-like, not DCE-like – they're generally designed to be used with a termination unit or modem, and their RS-232 ports are wired with DTE pinout. For now we need to keep the machine and device version of these things because we don't have a way to run two machines simultaneously, or to plug a device into multiple slots.
> 
> (The DCE-DCE bridge is to facilitate emulation of DCE-like things like serial printers, modems, etc. although page printers are another thing we need better support for.)

The DCE-DCE bridge actually fits perfectly here. Could I suggest another perspective. The emulator rs232 port bus has an interface and that interface is DCE like, and when connected to the terminal the terminal rs232 bus layer emulates the null modem cable between the terminal DTE interface and the rs232 port DTE interface. The interface is DCE like, it fits here.

This might also address the issues around the 'reset' method pushing values, rather than just passing them through, it is pushing values that are part of an active component, the null modem cable. Usually a null modem cable will pass through lines or return lines rather than actively driving them, but you could well have an active null modem cable, it is not necessarily wrong.

Unfortunately I squashed away the stand alone qvt102 front end, and will need a will need a little time to redo that, but will get to it if that is really the path people want. I thought you had a good plan to avoid a lot of redundant front ends here, sorry. Might you wish to reconsider relying on the DCE-DCE bridge, in light of it actually being a good fit here?

And thank you for checking the schematics, I was not aware there were any.

I made a quick attempt to define an interface class for the terminals to connect to the RS232 port bus. It was trivial to define methods for lines pushing values to the terminal and default methods for these, but it was not clear to me what pattern to use for the callback handlers. Could I ask if there is an existing pattern used in MAME that might be followed here, an interface class that is not a subclass of a device that implements callbacks?

smf- commented:
> This might also address the issues around the 'reset' method pushing values, rather than just passing them through, it is pushing values that are part of an active component, the null modem cable. Usually a null modem cable will pass through lines or return lines rather than actively driving them, but you could well have an active null modem cable, it is not necessarily wrong.

Null modem cables by definition just pass through values from one end to the other.
The null means there is no active logic. Therefore you can't have an active null modem cable. 

rs232 ports don't have a reset pin in any case, so even if it did exist then it would have no business implementing reset itself anyway.

68bit commented:
> > This might also address the issues around the 'reset' method pushing values, rather than just passing them through, it is pushing values that are part of an active component, the null modem cable. Usually a null modem cable will pass through lines or return lines rather than actively driving them, but you could well have an active null modem cable, it is not necessarily wrong.
> 
> Null modem cables by definition just pass through values from one end to the other.
> The null means there is no active logic. Therefore you can't have an active null modem cable.

Null modem cables are used to connect two DTE RS232 ports together. The 'null modem' means there are no modems between them, not that there is no active logic, unless by your definition an active cable is a modem but then all I can suggest is that that interpretation might not be helpful.

> rs232 ports don't have a reset pin in any case, so even if it did exist then it would have no business implementing reset itself anyway.

The DCD and DSR here a pushing constant values, as if they were driven by +12V. That these need to push these lines on an emulator reset is just an artifact of the emulator not pulling these values here.

Please have a look at the recently added dcebridge, by the very reviewer of this PR, and note that it pushes values for DTR and RTS on a reset when these are configured to 'assert' or 'deassert' these lines - that is all that is happening here just for different lines. Would you argue that that code was wrong for asserting active line levels?

I appreciate the review feedback, it is helpful, and I do lack experience on this code base. Actually I've been going back over my prior contributions and incorporating recent feedback to clean them up.

Perhaps I am still missing some important point and might need help from someone to explain it, sorry. Or perhaps when I take another look at it then it will occur to me. I have tired to describe what is on my mind, does that help at all?

smf- commented:
null modem cables are also called cross over cables, if that helps you. They are just cables. They don't contain anything that can generate a +/- 12v.

Does the code need to go in reset? Did you put it there because of ordering issues at startup?

68bit commented:
In the absence of any support for using the dcebridge to avoid redundant front ends the stand alone qvt102 front end has been retained in this PR.

> null modem cables are also called cross over cables, if that helps you. They are just cables. They don't contain anything that can generate a +/- 12v.

With respect, I don't think that position is useful. As noted an active connection can be trivially made that is powered and drives these lines. There are lots of serial lines that are active, some transition to optical fiber and back, some have repeaters, etc. Perhaps you would claim these were actually modems, but in that case this PR implements a modem connect to the terminal.

Not withstanding some disagreement over the definition of 'cable' or disagreement over the use of an active cable here, one recent precedent by the very reviewer of this PR has been noted that actively drives these lines from the `reset` method in particular the dcebridge. For another example please see the `null_modem.cpp` which does just the same, and the relevant source code from that file is noted below, doing exactly the same thing as in this proposed PR:

```
	output_rxd(1);

	// TODO: make this configurable
	output_dcd(0);
	output_dsr(0);
	output_cts(0);
```

Please this PR is not the right place to be debating this issue because it is a pre-existing strategy. I really don't care how it is eventually implemented so long as it is workable, but to address the matter will touch a good deal of code.

cuavas commented:
The long-term plan is to allow multiple machines to be run simultaneously, and have devices that can plug into multiple slots.  This will allow DTE-like things to be connected together.  MAME just isn't ready for it yet, so we have things like the machine/bus device duality for terminals.

68bit commented:
Rebased. The 'drive by' issues raised have landed, thanks. That still leaves A) the RS232 lines asserted in the reset method; B) that this might not be worth doing, it might be a dead end, and it might be better waiting until multiple machines can be connect in one instance.

A) On the issue of the RS232 lines being pushed in the reset method, it has not yet occurred to me what is claimed to be wrong there. I have tried to make the case above, and will not repeat it in full. I'll leave it with an attempted refutation to the claim that "this is a pass-through device and shouldn't generate values on its own". The QVT102 terminal RS232 port is a DTE like port and the CTS, DCD and DSR pins on that QVT102 port are inputs, not outputs, so it is not tenable to claim that these can "pass-through" to the PCs RS232 port inputs. Yes, I understand that there are a wide variety of null modem cables, some that cross over some of these lines, and some that loop back some of these lines, but those are not the particular connections proposed here. I think it would be useful to refactor the interface to have a more general null modem implementation, but that is not being proposed here, that is future work, what is proposed here is sufficient for this particular terminal.

B) It is acknowledged that this abstraction might be a dead end, that people might be undoing this patch in future when multiple machines can be connected together, and that might be a good reason to not land this PR, but then the IE15 and 8212 are in the same boat and probably need to be changed too. 

In any case I have a patch that I can apply locally while this issue is explored, and it is a nice improvement to be able to run this terminal along with some of these old machines, it allows them to run screen editors etc, and this works on the web. Thanks.

I don't wish to be involved in difficult issues, don't want to get into repetitive discussions, it's not my place here, likely I am just missing something, and I have other contributions to work on and propose and hope people will consider positively, so perhaps someone else would like to take up this one if there is any interest, or could I suggest we give up on it for now. Thanks again.

cuavas commented:
OK, let’s back up a bit.  MAME's soft reset (mapped to F3 by default) is supposed to roughly simulate what happens if you assert the system reset line.  MAME’s device reset is supposed to roughly simulate what happens when you assert a device's reset input.  There are certain limitations to the way it works that we'll get out of the way up-front:
* MAME assumes that resetting a device also resets all children.  This isn't necessarily true in real life – for example a hypothetical Virtual Boy cartridge containing a microcontroller may not wire the reset I/O line to the microcontroller's reset input.
* MAME assumes that resetting the system resets all devices in the system (this is an extension of the previous assumption, because soft reset is implemented by resetting the root device).  This is definitely not true in real life.  Sticking with Nintendo handhelds, there’s no reset connection from the system the system to the cartridge on the Game Boy or Game Boy Advance, so resetting the system won’t reset any devices in the cartridge (e.g. the smart controller for a multi-game Flash cartridge); on a PC, resetting the system won’t reset most external peripherals because there’s no system reset connection across serial ports, parallel ports, etc.
* MAME assumes that scheduled devices (CPUs, microcontrollers, etc.) will be in a static state as long as their reset input is asserted.  This isn’t always true in real life.  For example the 4004 and 4040 immediately zero all static registers when the reset input is asserted, but they dynamic registers are only zeroed when the refresh counter passes them while the reset input is asserted.

Device reset and soft reset aren’t supposed to produce a “clean start” state.  That’s achieved with the hard reset (Shift-F3 by default) which is implemented by tearing down the emulation session and starting again with the same configuration.

Some of MAME’s devices aren’t really devices in the sense of integrated circuits, they’re glue to implement a kind of arrangement of connections.  The input merger (src/devices/machine/input_merger.{h,cpp}) is a good example of this – it's used to represent multiple devices with open collector/open drain outputs connected to a single buffered input.  It provides a synchronisation barrier because it’s often used for communication between different executable devices.  Note that it *does not* implement `device_reset` because it just represents a particular connection arrangement.  The input seen by the buffer won’t change directly as a result of the system being reset.  The devices connected to it may change the state of their outputs when they’re reset, potentially causing the the output to change.  However the input merger device itself doesn’t do anything directly as a result of a reset.

The RS-232 bus devices for using terminals are conceptually similar.  They’re a null modem connected to a particular terminal, allowing it to be connected to a DCE.  They’re just a collection of wires.  In real life, there’s no button that resets both the terminal and the host (unless you do some serious wire mods), but MAME’s soft reset pretends there is.  Now when the devices on both sides get reset, they may change the state of their RS-232 outputs.  However the null modem itself *does not* do anything directly as a result of the reset.  It’s just a collection of wires.  It definitely shouldn’t be pushing a fixed RxD state to the host.  If resetting the terminal causes it to change its RxD output, that will should be passed through when it happens.

There are systems that depend on signals not getting magically changed out from under them when devices are reset (one example off the top of my head is some systems that expect EEPROM chip select to remain asserted across a reset).  This may not be one of them, but it’s still good practice to do things correctly within the limitations of MAME’s current framework.

68bit commented:
No support.

MooglyGuy commented:
Looking at the number of dev comments, you have received plenty of support, just not the kind of support you wanted. Now, when asked to actually improve an FDC controller rather than exacerbating MAME's problem of decades of legacy code that needs excision from the codebase, you're closing all of your pull requests in a move more resembling a toddler throwing toys from its pram than a functioning adult. Shame on you.

happppp commented:
He was very upset at the revert where hardly any explanation for it was given, this is not an unnatural reaction.

The legacy floppy device. It's comparable to let's say, someone adding sound support to a game using the legacy discrete sound system instead of netlist. Adding it makes it just a burden for future-us.

68bit commented:
I am not 'very upset' rather I understand why, but still need to change direction.

There was some constructive assistance with this PR and it was improved and some of the improvements have been committed separately, thanks. However it has been expressed that the MAME projected wanted to take a different path than moving a lot of terminals to the rs232 bus, and absolutely no one expressed support otherwise, see "The long-term plan is to allow multiple machines to be run simultaneously, and have devices that can plug into multiple slots." and that sounds great and I can understand people not wanting to move a lot of generic terminals to the rs232 bus if another approach is expected. I have a working patch for now, and improved thanks to the feedback, and that's something to be thankful for, thanks.

I don't want to be involved in disputes, and I don't believe I have done anything to invite these personal attacks. Someone else is welcome to look into it and address the claimed short comings or to commit the PR as is and then address the issues etc, and that might be helpful. I can also understand the frustration of experienced developers having to review low quality contributions, but there is only so much I can try to do, sorry, and I don't want to burden people.

Regarding the legacy floppy issue, the EXORset 30 patch will need very signification extra work, a rewrite of the patch so I chose to close it for now. It's no small block of work for me to rewrite the 6843 to use the modern floppy format, and I may just break existing systems causing more frustration. As I indicated another path is to get the EXORciser 6809 running and then the FDC that it uses with a 6809 ROM might be used for the EXORset 30, but there is a lot to explore there and the EXORset 30 standard configuration used the 6843 FDC for the 5.25 inch drives.

I understand that the MAME project might not want more legacy floppy code. It already appears a very large challenge moving all the current code, and imgtool etc, and the very many systems, to use the new floppy format. It looked like such a large challenge to me that it would never be done, but I don't want to be involved in such decisions, I'll just go with the flow.

rb6502 commented:
In the case where legacy systems require or are best used with a specific terminal I don't have a problem with making the terminal a device and having it available both via rs232 and standalone.

Legacy floppy is a harder problem, although a lot of systems have been moved off of it now (anything with an NEC765 or WD17XX controller, and Apple stuff is in progress).
